libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+
+		auto nodes = splitFlow<2>();
+
+		auto& whileBody = nodes[0];
+		auto& afterWhile = nodes[1];
+
+		m_currentNode = whileBody;
+		{
+			BreakContinueScope scope(*this, afterWhile, whileCondition);
+			appendControlFlow(_whileStatement.body());
+		}
+
+		connect(m_currentNode, whileCondition);
+
+		m_currentNode = afterWhile;
+	}
+
+
+	return false;
+}
+
+bool ControlFlowParser::visit(Break const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_breakJump, "");
+	connect(m_currentNode, m_breakJump);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Continue const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_continueJump, "");
+	connect(m_currentNode, m_continueJump);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Throw const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->revert, "");
+	connect(m_currentNode, m_cfg.m_currentFunctionFlow->revert);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+	return true;
+}
+
+void ControlFlowParser::endVisit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+}
+
+bool ControlFlowParser::visit(Return const& _return)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->exit, "");
+	// only the first return statement is interesting
+	if (!m_currentNode->block.returnStatement)
+		m_currentNode->block.returnStatement = &_return;
+	connect(m_currentNode, m_cfg.m_currentFunctionFlow->exit);
+	m_currentNode = newLabel();
+	return true;
+}
+
+
+bool ControlFlowParser::visit(PlaceholderStatement const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentModifierFlow, "");
+
+	connect(m_currentNode, m_cfg.m_currentModifierFlow->placeholderEntry);
+
+	m_currentNode = newLabel();
+
+	connect(m_cfg.m_currentModifierFlow->placeholderExit, m_currentNode);
+	return false;
+}
+
+bool ControlFlowParser::visitNode(ASTNode const& node)
+{
+	solAssert(!!m_currentNode, "");
+	if (auto const* expression = dynamic_cast<Expression const*>(&node))
+		m_currentNode->block.expressions.emplace_back(expression);
+	if (auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(&node))
+		m_currentNode->block.variableDeclarations.emplace_back(variableDeclaration);
+	if (auto const* assembly = dynamic_cast<InlineAssembly const*>(&node))
+		m_currentNode->block.inlineAssemblyStatements.emplace_back(assembly);
+
+	return true;
+}
+
+bool ControlFlowParser::visit(FunctionCall const& _functionCall)
+{
+	solAssert(!!m_currentNode, "");
+	if (auto functionType = dynamic_pointer_cast<FunctionType const>(_functionCall.expression().annotation().type))
+		switch (functionType->kind())
+		{
+			case FunctionType::Kind::Revert:
```

Comment:
Can you still visit the expression in case of revert, require and assert? We might support weird stuff in the future, just to make sure.